client/allocrunner/taskrunner/connect_hook.go

```diff
@@ -0,0 +1,145 @@
+package taskrunner
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/allocdir"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	agentconsul "github.com/hashicorp/nomad/command/agent/consul"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+var _ interfaces.TaskPrestartHook = &connectHook{}
+
+// connectHook writes the bootstrap config for the Connect proxy sidecar
+type connectHook struct {
+	alloc          *structs.Allocation
+	consulHTTPAddr string
+
+	logger log.Logger
+}
+
+func newConnectHook(alloc *structs.Allocation, consulHTTPAddr string, logger log.Logger) *connectHook {
+	h := &connectHook{
+		alloc:          alloc,
+		consulHTTPAddr: consulHTTPAddr,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (connectHook) Name() string {
+	return "connect"
+}
+
+func (h *connectHook) Prestart(ctx context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	if !req.Task.Kind.IsConnectProxy() {
+		// Not a Connect proxy sidecar
+		resp.Done = true
+		return nil
+	}
+
+	serviceName := req.Task.Kind.Value()
+	if serviceName == "" {
+		return fmt.Errorf("Connect proxy sidecar does not specify service name")
+	}
+
+	tg := h.alloc.Job.LookupTaskGroup(h.alloc.TaskGroup)
+
+	var service *structs.Service
+	for _, s := range tg.Services {
+		if s.Name == serviceName {
+			service = s
+		}
+	}
+
+	if service == nil {
+		return fmt.Errorf("Connect proxy sidecar task exists but no services configured with a sidecar")
+	}
+
+	h.logger.Debug("bootstrapping Connect proxy sidecar", "task", req.Task.Name, "service", serviceName)
+
+	canary := false
+	if h.alloc.DeploymentStatus != nil {
+		canary = h.alloc.DeploymentStatus.Canary
+	}
+
+	//TODO(schmichael) relies on GRPCSocket being created
+	//TODO(schmichael) unnecessasry if the sidecar is running on the host netns
+	grpcAddr := "unix://" + filepath.Join(allocdir.SharedAllocName, allocdir.AllocGRPCSocket)
+	fn := filepath.Join(req.TaskDir.LocalDir, "envoy_bootstrap.json")
+	id := agentconsul.MakeTaskServiceID(h.alloc.ID, "group-"+tg.Name, service, canary)
+	h.logger.Debug("bootstrapping envoy", "sidecar_for", service.Name, "boostrap_file", fn, "sidecar_for_id", id, "grpc_addr", grpcAddr)
+
+	// Since Consul services are registered asynchronously with this task
+	// hook running, retry a small number of times with backoff.
+	for tries := 3; ; tries-- {
+		cmd := exec.CommandContext(ctx, "consul", "connect", "envoy",
+			"-grpc-addr", grpcAddr,
+			"-http-addr", h.consulHTTPAddr,
+			"-bootstrap",
+			"-sidecar-for", id,
+		)
+
+		// Redirect output to local/bootstrap.json
+		fd, err := os.Create(fn)
```

Comment:
Should we set any particular permissions on this file? (I think it's 0666 before umask by default.)

Reply:
I feel like you deserve a bonus just for asking. :sweat_smile: 

The bootstrap file may contain secrets such as a Consul token. For this reason I think it's better to place in [`${NOMAD_SECRETS_DIR}`](https://www.nomadproject.io/docs/runtime/environment.html#task-directories) than to try to use Unix permissions to protect. There's no per-task user, so making the file `0400` and owned by `nobody` doesn't really secure it any more than placing it in `secrets/` does.

We do offer more protections for `secrets/` than other directories (eg it can't be read remotely via the filesystem APIs without a special ACL).